warehouse/query-core/src/main/java/datawave/query/util/transformer/AttributeRebuilder.java

```diff
@@ -0,0 +1,123 @@
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.model.QueryModel;
+import datawave.query.util.TypeMetadata;
+
+/**
+ * Utility class that handles rebuilding a basic NoOp attribute into an attribute of the correct type, given TypeMetadata and optionally a QueryModel
+ */
+public class AttributeRebuilder {
+    private static final Logger log = Logger.getLogger(AttributeRebuilder.class);
+    private final TypeMetadata typeMetadata;
+    private final QueryModel queryModel;
+    private final Map<String,Class<?>> classCache;
+
+    /**
+     * Main constructor for the rebuilder, QueryModel may be null
+     *
+     * @param typeMetadata
+     * @param queryModel
+     */
+    public AttributeRebuilder(TypeMetadata typeMetadata, @Nullable QueryModel queryModel) {
+        this.typeMetadata = typeMetadata;
+        this.queryModel = queryModel;
+        this.classCache = new HashMap<>();
+    }
+
+    /**
+     * Given a field and an attribute, return the correctly typed attribute
+     *
+     * @param field
+     *            the field
+     * @param attr
+     *            the attribute
+     * @return an attribute of the correct type
+     */
+    public Attribute rebuild(String field, Attribute<?> attr) {
+        field = JexlASTHelper.deconstructIdentifier(field);
+        Set<String> normalizerNames = typeMetadata.getNormalizerNamesForField(field);
+
+        if (normalizerNames.isEmpty()) {
+            populateNormalizerFromQueryModel(field, normalizerNames);
+        }
+
+        if (normalizerNames.size() > 1 && log.isTraceEnabled()) {
+            log.trace("Found " + normalizerNames.size() + " normalizers for field " + field + ", using first normalizer");
+        }
+
+        for (String name : normalizerNames) {
+            try {
+                Class<?> clazz = getClass(name);
+                Type<?> type = (Type<?>) clazz.getDeclaredConstructor().newInstance();
+
+                type.setDelegateFromString(String.valueOf(attr.getData()));
+                return new TypeAttribute<>(type, attr.getMetadata(), attr.isToKeep());
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        return attr;
+    }
+
+    /**
+     * Populate normalizer names from the query model
+     *
+     * @param field
+     *            the field
+     * @param normalizerNames
+     *            the set of normalizer names
+     */
+    private void populateNormalizerFromQueryModel(String field, Set<String> normalizerNames) {
+        if (log.isTraceEnabled()) {
+            log.trace("Field " + field + " not found in TypeMetadata, falling back to QueryModel");
+        }
+
+        if (queryModel == null) {
+            log.warn("QueryModel is null, cannot populate normalizers for " + field + " from model");
+            return;
+        }
+
+        // check forward mappings
+        Multimap<String,String> forwardMappings = queryModel.getForwardQueryMapping();
```

Comment:
I think you may be going the wrong way on the mappings here.  At the point where you are rebuilding attributes, if you have the fields as they exist in the DB then there is no mapping required.  If you have the attributes after applying the reverse mappings, then you would nee the reverse mappings here.  In any case I cannot see how the forward mappings would be applicable here.